### ISSUE TITLE
Icompatability with other modules adding event listeners in "renderJournalSheet" Hook

### ISSUE THREAD

**Issue by Evannessa** [Miskatonic-Investigative-Society/CoC7-FoundryVTT#325]
Foundry Version 0.7.9
CoC7 - System Version 0.4.7

Hey there! I'm the developer of the [journal-to-canvas-slideshow.](https://github.com/EvanesceExotica/Journal-To-Canvas-Slideshow) module. I ran into an issue earlier with a user not being able to properly use my module, and we pinpointed the issue to this system, as it worked as expected in other systems.

(As a forewarning, I'm fairly new to JavaScript, so some of my code might be bad practice and I don't quite understand everything.)

Specifically, I _think_ the problem might lie in the _ParseSheetContent_ and _EnrichHTML_ methods of the "_parser.js_" script, where in the _EnrichHTML_ method, I think the journal entry's content is being set using the ._innerHTML_ property, and that was removing the event listeners I'd set with my module's "_renderJournalSheet_" hook. (My code sets event listeners for "click" and "mouseover" events in the "_renderJournalSheet_" hook. ) 

I think I found a way around it by using Promises to 'delay' my code's event listeners being set, and therefore I can probably more easily secure a solution on my end, but I wanted to make you aware just in case there are other modules that end up having trouble with this as well. Maybe something like .insertAdjacentHTML() could be used in place of .innerHTML? 

Again, I'm pretty new to development so please forgive any misunderstandings. But just wanted to make you aware. ^ ^

**Comment by HavlockV** [comment]
Thanks for the feedback, I'll have a look at it ASAP

**Comment by kckaiwei** [comment]
@EvanesceExotica I want to add on here that it seems the current implementation for CoC7 is replicating how foundry is enriching the text:

https://foundryvtt.com/api/foundry.js.html#line13462

It seems that HavlockV basically changed up how foundry was originally enriching the text, and used the same method, so it's very likely that other modules who also copied FoundryVTT's method will run into the issue you mentioned.

I'm not fully familiar with CoC7 since I only started reading it today, but I'm not sure that .insertAdjacentHTML() would work in this case, since I think it's parsing for keywords, and then replacing the whole message, which I don't think would be possible with .insertAdjacentHTML()?

**Comment by HavlockV** [comment]
Hi, so I checked.
Great module by the way, very handy !
I wasn't able to reproduce the bug. Created a display scene, a journal entry with images and a CoC7 link and it's working, both the link and the images.
Could you provide more background (maybe a copy of a non working journal entry ?) so I can investigate further ?

**Comment by HavlockV** [comment]
I didn't see any reason why your event listener would be overiden. 
In doubt I made a quick 'fix'.
Before, every element was parsed and replaced even if that element didn't contain any CoC7 link.
Now an element will be parsed and replaced only if he contains a link (anything starting with @CoC7).
It's on the develop branch. Clone it and see if it solves the problem.

**Comment by Evannessa** [comment]
@HavlockV 
Apologies for the late reply! I recently found a quick fix solution on my end and pushed it to GitHub/updated my module, so it should be working properly now, and the user just confirmed it was working for them as well. You may have been trying it out with the updated version -- sorry for not getting back to you sooner.

I basically changed my event listeners from being added in the "renderJournalSheet" hook in my ClickImageInJournal.js file to being added afterward using a Promise and the .then() method, so I think it avoids what was happening before with the event listeners getting overriden. 

My original version was like this:
```
Hooks.on("renderJournalSheet", (app, html, options) => {
	html.find('img').attr("class", "clickableImage");
	html.find('video').attr("class", "clickableImage");
	html.find(".lightbox-image").each((i, div) => {
		div.classList.add("clickableImage");
	})

	html.find('.clickableImage').each((i, div) => {
		div.addEventListener("click", displayImage, false);
		div.addEventListener("mouseover", highlight, false);
		div.addEventListener("mouseout", dehighlight, false);
		div.addEventListener("mousedown", depressImage, false);
		div.addEventListener("mouseup", liftImage, false);
	});


});
```

But I've changed it to this for now as a quick fix, and it seemed to solve the problem on my end:

```
function setEventListeners(html) {
	wait().then(execute.bind(null, html));
}

function wait(callback) {
	return new Promise(function (resolve, reject) {
		resolve();
	})
}

function execute(html) {
	html.find('.clickableImage').each((i, div) => {
		div.addEventListener("click", displayImage, false);
		div.addEventListener("mouseover", highlight, false);
		div.addEventListener("mouseout", dehighlight, false);
		div.addEventListener("mousedown", depressImage, false);
		div.addEventListener("mouseup", liftImage, false);
	});
}

Hooks.on("renderJournalSheet", (app, html, options) => {
	html.find('img').attr("class", "clickableImage");
	html.find('video').attr("class", "clickableImage");
	html.find(".lightbox-image").each((i, div) => {
		div.classList.add("clickableImage");
	})

	setEventListeners(html);
	
});
```
I'll clone and take a look at what you mentioned on the development branch. Thanks for the help!

**Comment by Evannessa** [comment]
@kckaiwei 

> It seems that HavlockV basically changed up how foundry was originally enriching the text, and used the same method, so it's very likely that other modules who also copied FoundryVTT's method will run into the issue you mentioned.

> I'm not fully familiar with CoC7 since I only started reading it today, but I'm not sure that .insertAdjacentHTML() would work in this case, since I think it's parsing for keywords, and then replacing the whole message, which I don't think would be possible with .insertAdjacentHTML()?



Ah, ok! That makes sense. I'll try to look into it further. As mentioned, I'm pretty new to all of this, so thanks so much for the information ^ ^

**Comment by HavlockV** [comment]
Investigated a little further. As mentioned before, the parser was scanning for all <p> elements in all editors.
The content of every <p> was replaced regardless of that element actually containing a CoC7 link.
In the process all events attached to any child element was removed.
( in : `<p><img src="1.png" alt="" width="123" height="123" /></p>` all eventHandler attached to img was overwritten)
I added a test to replace only elements containing a CoC7 link, not replacing other elements.
I think the promise was just making sure that the images were parsed and eventHandler attached after the links.
The promise shouldn't be necessary anymore.
As pointed out bu @kckaiwei Foundry uses the same thing when parsing editors contents (to replace the inline rolls for example).
I ran on a similar issue developing this, I think it was with PDFoudry module (not sure).
I ended up binding my events not on individual elements but on the parent html of every sheet, to prevent the element handler to be removed.
